### DISCUSSION TITLE
Lesson 11: Cannot import LinkToken from Brownie...

### DISCUSSION THREAD

**Discussion by toddabraham** [smartcontractkit/full-blockchain-solidity-course-py#129] [Q&A]
i'm trying to run my `deploy_and_create.py` and I get this error (occurs for both Simple and Advanced)....

`ImportError: cannot import name 'LinkToken' from 'brownie'`

`todd@fcc-blockchain-dev:~/fcc_blockchain_dev/nft-demo$ brownie run scripts/advanced_collectible/deploy_and_create.py 
Brownie v1.16.3 - Python development framework for Ethereum

NftDemoProject is the active project.

Launching 'ganache-cli --accounts 10 --hardfork istanbul --gasLimit 12000000 --mnemonic brownie --port 8545'...
  File "brownie/_cli/run.py", line 49, in main
    return_value, frame = run(
  File "brownie/project/scripts.py", line 53, in run
    module = _import_from_path(script)
  File "brownie/project/scripts.py", line 149, in _import_from_path
    _import_cache[import_str] = importlib.import_module(import_str)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen, line line, in in
  File "<frozen, line line, in in
  File "<frozen, line line, in in
  File "<frozen, line line, in in
  File "<frozen, line line, in in
  File "<frozen, line line, in in
  File "./scripts/advanced_collectible/deploy_and_create.py", line 1, in <module>
    from scripts.helpful_scripts import (
  File "./scripts/helpful_scripts.py", line 1, in <module>
    from brownie import accounts, network, config, LinkToken, VRFCoordinatorMock, Contract
ImportError: cannot import name 'LinkToken' from 'brownie' (/home/todd/.local/pipx/venvs/eth-brownie/lib/python3.8/site-packages/brownie/__init__.py)
Terminating local RPC client...`

**Comment by PatrickAlphaC** [comment]
Wow lesson 11!! Great work so far. Could you put your deploy script here?

**Reply by toddabraham** [reply]
Sure thing, here you go:

`deploy_and_create.py`

```python
from scripts.helpful_scripts import (
    get_account,
    OPENSEA_URL,
    get_contract,
    fund_with_link,
)
from brownie import AdvancedCollectible, network, config


def deploy_and_create():
    account = get_account()
    advanced_collectible = AdvancedCollectible.deploy(
        get_contract("vrf_coordinator"),
        get_contract("link_token"),
        config["networks"][network.show_active()]["keyhash"],
        config["networks"][network.show_active()]["fee"],
        {"from": account},
    )
    fund_with_link(advanced_collectible.address)
    creating_tx = advanced_collectible.createCollectible({"from": account})
    creating_tx.wait(1)
    print("New token has been created!")
    return advanced_collectible, creating_tx


def main():
    deploy_and_create()
```

**Reply by toddabraham** [reply]
`helpful_scripts.py`

```python
from brownie import accounts, network, config, LinkToken, VRFCoordinatorMock, Contract
from web3 import Web3

LOCAL_BLOCKCHAIN_ENVIRONMENTS = ["hardhat", "development", "ganache", "mainnet-fork"]
OPENSEA_URL = "https://testnets.opensea.io/assets/{}/{}"
BREED_MAPPING = {0: "PUG", 1: "SHIBA_INU", 2: "ST_BERNARD"}


def get_breed(breed_number):
    return BREED_MAPPING[breed_number]


def get_account(index=None, id=None):
    if index:
        return accounts[index]
    if network.show_active() in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        return accounts[0]
    if id:
        return accounts.load(id)
    return accounts.add(config["wallets"]["from_key"])


contract_to_mock = {"link_token": LinkToken, "vrf_coordinator": VRFCoordinatorMock}


def get_contract(contract_name):
    """
    This function will either:
        - Get an address from the config
        - Or deploy a Mock to use for a network that doesn't have the conract
    Args:
        contract_name (string): This is the name of the contract that we will get
        from the config or deploy
    Returns:
        brownie.network.contract.ProjectContract: This is the most recently deployed
        Contract of the type specified by a dictionary. This could either be a mock
        or a 'real' contract on a live network.
    """
    # link_token
    # LinkToken
    contract_type = contract_to_mock[contract_name]
    if network.show_active() in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        if len(contract_type) <= 0:
            deploy_mocks()
        contract = contract_type[-1]
    else:
        contract_address = config["networks"][network.show_active()][contract_name]
        contract = Contract.from_abi(
            contract_type._name, contract_address, contract_type.abi
        )
    return contract


def deploy_mocks():
    """
    Use this script if you want to deploy mocks to a testnet
    """
    print(f"The active network is {network.show_active()}")
    print("Deploying mocks...")
    account = get_account()
    print("Deploying Mock LinkToken...")
    link_token = LinkToken.deploy({"from": account})
    print(f"Link Token deployed to {link_token.address}")
    print("Deploying Mock VRF Coordinator...")
    vrf_coordinator = VRFCoordinatorMock.deploy(link_token.address, {"from": account})
    print(f"VRFCoordinator deployed to {vrf_coordinator.address}")
    print("All done!")


def fund_with_link(
    contract_address, account=None, link_token=None, amount=Web3.toWei(0.3, "ether")
):
    account = account if account else get_account()
    link_token = link_token if link_token else get_contract("link_token")
    funding_tx = link_token.transfer(contract_address, amount, {"from": account})
    funding_tx.wait(1)
    print(f"Funded {contract_address}")
    return funding_tx
```

**Reply by toddabraham** [reply]
The error is now changed to VRFCoordinatorMock.

![image](https://user-images.githubusercontent.com/26919212/135763121-6c131b13-3073-4e14-bcaf-420a9ef9d5b6.png)


I assume VRFCoordinatorMock should be showing up in my AdvancedCollectible.json file too, but it's not....

![image](https://user-images.githubusercontent.com/26919212/135763057-e147c1ff-424e-44db-a4c8-3be73a1d3ef8.png)

**Reply by toddabraham** [reply]
Tried uploading AdvancedCollectible.json but it's way too big.

**Reply by PatrickAlphaC** [reply]
Thanks for the information. It looks like you might be missing the contracts in your `contracts/test` folder. Can you show what's in your `contracts/test` or `contracts/tests` folder?

**Reply by toddabraham** [reply]
@PatrickAlphaC Not sure how I missed that.  Here's what I have....

![image](https://user-images.githubusercontent.com/26919212/136142696-6cdb49a1-6a06-4417-9680-ecbd78a64305.png)

**Reply by PatrickAlphaC** [reply]
Almost, I meant in your `contracts/tests` folder. Not your `tests` folder

**Reply by toddabraham** [reply]
@PatrickAlphaC I haven't yet created those files.  Is that what at video TS 10:46:04 you say to "look in the Github repo for a couple try catches there that make it a little nicer for error handling?"

![image](https://user-images.githubusercontent.com/26919212/136243766-c9492db5-a930-410f-a1e3-9e2b3100e6dc.png)

**Reply by toddabraham** [reply]
@PatrickAlphaC Ok, I added a tests folder inside my contracts folder and copied in LinkToken.sol and VRFCoordinatorMock.sol, but I'm still getting the error....perplexing!

![image](https://user-images.githubusercontent.com/26919212/136246519-47e6411f-24ac-432f-89d3-d845359e28b8.png)

**Reply by toddabraham** [reply]
@PatrickAlphaC Fixed.  Had /tests instead of /test

![image](https://user-images.githubusercontent.com/26919212/136248171-da5fa06f-ab45-4d0c-86b4-94f1e59a0393.png)

**Reply by toddabraham** [reply]
@PatrickAlphaC As a suggestion, to help learners as slow as I am, maybe add a little more explanation on the test files LinkToken.sol and VRFCoordinatorMock.sol.  I ended up going back over the video trying to find my error but didn't remember to add them as in the lottery project.

**Reply by PatrickAlphaC** [reply]
Hmm... This is good feedback. Thanks @toddabraham! Love how much you're learning!!

**Reply by mateuscgodoy** [reply]
I would like to reinforce Todd's suggestion! Got caught by the same rookie mistake. 
But anyways, thank you for all the support and high quality knowledge so far!

**Reply by WeTheArtist** [reply]
Thank you for this, while reading this I was able to understand why my code was not working ok. Although it was a completely different error. :)
